src/runtime_src/core/common/api/xrt_module.cpp

```diff
@@ -425,27 +429,82 @@ generate_key_string(const std::string& argument_name, xrt_core::patcher::buf_typ
   return argument_name + buf_string;
 }
 
+
+// Basic Itanium ABI type decoding. get_demangled_type() is referenced from ChatGPT response
 static std::string
-demangle(const std::string& mangled_name)
+get_demangled_type(const std::string& s, size_t& idx)
 {
-#ifdef _WIN32
-  char demangled_name[1024];
-  if (UnDecorateSymbolName(mangled_name.c_str(), demangled_name, sizeof(demangled_name), UNDNAME_COMPLETE))
-    return std::string(demangled_name);
-  else
-    throw std::runtime_error("Error demangling kernel signature");
-#else
-  int status = 0;
-  std::unique_ptr<char, decltype(&std::free)> demangled_name
-    (abi::__cxa_demangle(mangled_name.c_str(), nullptr, nullptr, &status), std::free);
+  if (idx >= s.size())
+    throw std::runtime_error("demangle arg index out of bounds");
+
+  // map of mangled argument types to their demangled string representations
+  static const std::map<char, std::string> demangle_type_map = {
+    {'v', "void"},
+    {'c', "char"},
+    {'i', "int"}
+  };
 
-  if (status)
-    throw std::runtime_error("Error demangling kernel signature");
+  char c = s[idx++];
 
-  return demangled_name.get();
-#endif
+  // Append "*" for pointer types
+  if (c == 'P')
+    return get_demangled_type(s, idx) + "*";
+
+  auto it = demangle_type_map.find(c);
+  if (it == demangle_type_map.end())
+    throw std::runtime_error("Unknown type character in mangled name: " + std::string(1, c));
+
+  return it->second;
 }
 
+// Parse mangled name in Itanium ABI style: _Z<length><name><types>
+// length : number of characters in the name string.
+// name   : kernel name in string
+// types  : kernel argument data type as below.
+//  'c' represents the arg is a char.
+//  'v' represents the arg is a void.
+//  'i' represents the arg is an int.
+//  'P' represents the arg is a pointer.
+//      Hence, "Pc" = char*, "Pv" = void*, "Pi" = int*
+// demangle() is referenced from ChatGPT response
+static std::string
+demangle(const std::string& mangled)
+{
+  //Check if mangled prefix "_Z" is present and length is greater than mangled_prefix_length
+  if (mangled.size() <= mangled_prefix_length || mangled.substr(0, mangled_prefix_length) != "_Z")
+    throw std::runtime_error("Doesn't have prefix _Z, not a mangled kernel name");
+
+  size_t idx = 2;
+  size_t len = 0;
+
+  // Extract length of function name
+  while (idx < mangled.size() && std::isdigit(mangled[idx]))
+    len = len * decimal_base + (mangled[idx++] - '0');
+
+  if (idx + len > mangled.size())
+    throw std::runtime_error("Invalid mangled name, doesn't have expected kernel name length");
+
+  std::string name = mangled.substr(idx, len);
+  idx += len;
+  std::vector<std::string> args;
+
+  // Extract type of function arguments
+  while (idx < mangled.size())
+    args.push_back(get_demangled_type(mangled, idx));
```

Comment:
How about
```suggestion
  while (idx < mangled.size())
    args.push_back(get_demangled_type(mangled[idx++]));
```

or simpler maybe:

```
for (auto c : mangled)
  args.push_back(get_demangled_type(c));
```

Reply:
done